src/operations.jl

```diff
@@ -1,91 +1,107 @@
+  ctprod! = @closure (res, w, α, β) -> mul!(res, adjoint(op), w, -α, β)
+  LinearOperator{T}(op.nrow, op.ncol, op.symmetric, op.hermitian, prod!, tprod!, ctprod!)
 end
 
-function mul!(y::AbstractVector, op::AbstractLinearOperator, x::AbstractVector)
-  y .= op * x
-  return y
+function prod_op!(res::AbstractVector, op1::AbstractLinearOperator, op2::AbstractLinearOperator, 
+                  vtmp::AbstractVector, v::AbstractVector, α, β)
+  mul!(vtmp, op2, v)
+  mul!(res, op1, vtmp, α, β)
 end
 
-# Binary operations.
+function ctprod_op!(res::AbstractVector, op1::AbstractLinearOperator, op2::AbstractLinearOperator, 
+                    wtmp::AbstractVector, w::AbstractVector, α, β)
+  mul!(wtmp, op1, w)
+  mul!(res, op2, wtmp, α, β)
+end
 
 ## Operator times operator.
 function *(op1::AbstractLinearOperator, op2::AbstractLinearOperator)
+  T = promote_type(eltype(op1), eltype(op2))
   (m1, n1) = size(op1)
   (m2, n2) = size(op2)
   if m2 != n1
     throw(LinearOperatorException("shape mismatch"))
   end
-  S = promote_type(eltype(op1), eltype(op2))
-  prod = @closure v -> op1 * (op2 * v)
-  tprod = @closure u -> transpose(op2) * (transpose(op1) * u)
-  ctprod = @closure w -> op2' * (op1' * w)
-  LinearOperator{S}(m1, n2, false, false, prod, tprod, ctprod)
+  #tmp vector for products
+  vtmp = zeros(T, m2)
+  utmp = zeros(T, n1)
+  wtmp = zeros(T, n1)
+  prod! = @closure (res, v, α, β) -> prod_op!(res, op1, op2, vtmp, v, α, β)
+  tprod! = @closure (res, u, α, β) -> ctprod_op!(res, transpose(op1), transpose(op2), utmp, u, α, β)
+  ctprod! = @closure (res, w, α, β) -> ctprod_op!(res, adjoint(op1), adjoint(op2), wtmp, w, α, β)
+  LinearOperator{T}(m1, n2, false, false, prod!, tprod!, ctprod!)
 end
 
 ## Matrix times operator.
 *(M::AbstractMatrix, op::AbstractLinearOperator) = LinearOperator(M) * op
 *(op::AbstractLinearOperator, M::AbstractMatrix) = op * LinearOperator(M)
 
-## Scalar times operator.
+## Scalar times operator. (# commutation α*v ???)
 function *(op::AbstractLinearOperator, x::Number)
   S = promote_type(eltype(op), typeof(x))
-  prod = @closure v -> (op * v) * x
-  tprod = @closure u -> x * (transpose(op) * u)
-  ctprod = @closure w -> x' * (op' * w)
-  LinearOperator{S}(op.nrow, op.ncol, op.symmetric, op.hermitian && isreal(x), prod, tprod, ctprod)
+  prod! = @closure (res, v, α, β) -> mul!(res, op, v, x * α, β)
+  tprod! = @closure (res, u, α, β) -> mul!(res, transpose(op), u, x * α, β)
+  ctprod! = @closure (res, w, α, β) -> mul!(res, adjoint(op), w, x' * α, β)
+  LinearOperator{S}(op.nrow, op.ncol, op.symmetric, op.hermitian && isreal(x), prod!, tprod!, ctprod!)
 end
 
 function *(x::Number, op::AbstractLinearOperator)
-  S = promote_type(eltype(op), typeof(x))
-  prod = @closure v -> x * (op * v)
-  tprod = @closure u -> (transpose(op) * u) * x
-  ctprod = @closure w -> (op' * w) * x'
-  LinearOperator{S}(op.nrow, op.ncol, op.symmetric, op.hermitian && isreal(x), prod, tprod, ctprod)
+  return op * x
```

Comment:
We are assuming they commute here?!

Reply:
That's mandatory if `x` is a scalar no? The result of `op * v` was previously a vector so it does not change anything to left or right multiply, or am I missing something?